### PR TITLE
TF-2083 Add/remove conditions in UI

### PR DESCRIPTION
**Pull request by hieutbui**
## Issue
---
 #2083 

## Resolved
---

https://github.com/linagora/tmail-flutter/assets/80142234/8deca3b1-a814-405f-869d-71016f2e5467

### CODE REVIEW

File: lib/features/rules_filter_creator/presentation/model/rule_filter_condition_type.dart [hunk 1]
```diff
@@ -0,0 +1,5 @@
+enum RuleFilterConditionType {
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/model/rule_filter_condition_type.dart:1]:
```suggestion
enum RuleFilterConditionScreenType
```

**Review by Arsnael — COMMENTED** [review]
Thanks for tagging me, was interesting to read.

There is maybe a concern I have, but maybe I misunderstood something (in case don't hesitate to tell me so).

I feel like in the end, when creating multiple conditions on a rule, what you seem to send (but maybe I'm wrong) to the tmail backend is a list of rules corresponding to each condition, while actually the backend would I think expect getting only one rule, but with the multiple conditions. 

I would be curious to see a video where you click on the create button after adding a few conditions to your rules and see if you get a list of rules or just one rule, and then seeing if when you try to edit if everything gets back?

Also on the video: mobile version I did not see the removal button on conditions

Otherwise good job :)

**Comment by Arsnael** [comment]
Ah nvm I just reread the title, it's only for the UI so far correct, so the rest about the logic with the backend is gonna be in an other PR?

**Comment by hieutbui** [comment]
> Ah nvm I just reread the title, it's only for the UI so far correct, so the rest about the logic with the backend is gonna be in an other PR?

Yeah, it will be handled in the next PR

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart [hunk 2]
```diff
@@ -48,21 +50,19 @@ class RulesFilterCreatorController extends BaseMailboxController {
...
   final mailboxSelected = Rxn<PresentationMailbox>();
   final actionType = CreatorActionType.create.obs;
+  final listRuleCondition = RxList<RuleCondition>();
 
   final TextEditingController inputRuleNameController = TextEditingController();
-  final TextEditingController inputConditionValueController = TextEditingController();
+  final List<TextEditingController> listInputConditionValueController = [];
   final FocusNode inputRuleNameFocusNode = FocusNode();
-  final FocusNode inputRuleConditionFocusNode = FocusNode();
+  final List<FocusNode> listInputRuleConditionFocusNode = [];
```
Comment (drminh2807) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart:63]: listErrorRuleConditionValue, listInputConditionValueController, listErrorRuleConditionValue are same length,
so I think better is combine it into a class, then define a List for that class instance only

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart [hunk 3]
```diff
@@ -143,39 +146,57 @@ class RuleFilterCreatorView extends GetWidget<RulesFilterCreatorController> {
                        fontSize: 16,
                        color: Colors.black)),
                   const SizedBox(height: 24),
+                  Obx(() {
+                    return ListView.separated(
```
Comment (drminh2807) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_view.dart:150]: DRY, ListView.separated duplicate 3 times in this file

File: lib/features/rules_filter_creator/presentation/model/rules_filter_input_field_arguments.dart [hunk 4]
```diff
@@ -0,0 +1,22 @@
+// ignore_for_file: public_member_api_docs, sort_constructors_first
+import 'package:equatable/equatable.dart';
+import 'package:flutter/material.dart';
+
+class RulesFilterInputFieldArguments with EquatableMixin {
+  final FocusNode inputRuleConditionValueFocusNode;
```
Comment (drminh2807) [inline lib/features/rules_filter_creator/presentation/model/rules_filter_input_field_arguments.dart:6]:
```suggestion
  final FocusNode focusNode;
```
I think class name is enough to describe, so no need long param name

File: lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart [hunk 5]
```diff
@@ -39,6 +41,8 @@ import 'package:tmail_ui_user/main/routes/app_routes.dart';
 import 'package:tmail_ui_user/main/routes/dialog_router.dart';
 import 'package:tmail_ui_user/main/routes/route_navigation.dart';
 
+import 'model/rules_filter_input_field_arguments.dart';
```
Comment (drminh2807) [inline lib/features/rules_filter_creator/presentation/rules_filter_creator_controller.dart:44]: can you try import from package ? same with another import

File: lib/features/rules_filter_creator/presentation/model/rules_filter_input_field_arguments.dart [hunk 6]
```diff
@@ -0,0 +1,22 @@
+// ignore_for_file: public_member_api_docs, sort_constructors_first
+import 'package:equatable/equatable.dart';
+import 'package:flutter/material.dart';
+
+class RulesFilterInputFieldArguments with EquatableMixin {
+  final FocusNode fucusNode;
```
Comment (drminh2807) [inline lib/features/rules_filter_creator/presentation/model/rules_filter_input_field_arguments.dart:6]:
```suggestion
  final FocusNode focusNode;
```

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_condition_widget.dart [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+            );
+          }
+        );
+      })
+    );
+    
+  }
+}
+
+Widget _buildRuleFilterCondition(
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_condition_widget.dart:105]: Create Stateless widget for it

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_condition_widget.dart [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+              child: _buildRemoveRuleFilterConditionButton(tapRemoveRuleFilterConditionCallback, imagePaths)
+            ),
+          ],
+        );
+      default:
+        return Container();
+    }
+}
+
+Widget _buildRemoveRuleFilterConditionButton (
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_condition_widget.dart:198]: same

File: lib/features/rules_filter_creator/presentation/model/rules_filter_input_field_arguments.dart [hunk 9]
```diff
@@ -0,0 +1,22 @@
+// ignore_for_file: public_member_api_docs, sort_constructors_first
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/model/rules_filter_input_field_arguments.dart:1]: remove it

**Review by hoangdat — APPROVED** [review]

**Review by dab246 — APPROVED** [review]

**Pull request merged**